olive/model.py

```diff
@@ -58,21 +58,29 @@ def __init__(
         self.is_aml_model = is_aml_model
 
     @abstractmethod
-    def load_model(self) -> object:
+    def load_model(self, rank: int = None) -> object:
         """
         Load model from disk, return in-memory model object
         Derived class should implement its specific logic if needed.
         """
         raise NotImplementedError()
 
     @abstractmethod
-    def prepare_session(self, inference_settings: Optional[Dict[str, Any]] = None, device: Device = Device.CPU):
+    def prepare_session(self, inference_settings: Optional[Dict[str, Any]] = None,
+                        device: Device = Device.CPU, rank: int = None):
         """
         Prepare inference session for Olive model, return in-memory inference session.
         Derived class should implement its specific logic if needed.
         """
         raise NotImplementedError()
 
+    @abstractmethod
+    def get_default_execution_providers(self, device: Device):
```

Comment:
`execution_providers` is not suitable for torch model. Should be put it under this base class?

Reply:
Also, it leads a failure as there is no implementation in TorchModel for this `abstractmethod`.

Reply:
Reworked logic with the newly introduced ONNXModelBase